lrlex/src/lib/ctbuilder.rs

```diff
@@ -79,6 +80,26 @@ pub enum RustEdition {
     Rust2021,
 }
 
+pub trait LexErrorHandler<LexerTypesT>
+where
+    LexerTypesT: LexerTypes,
+    usize: num_traits::AsPrimitive<LexerTypesT::StorageT>,
+{
+    /// Called with the lexers filename
+    fn lexer_path(&mut self, filename: &Path);
+    /// Called with the lexers source contents.
+    fn lexer_src(&mut self, src: &str);
+    fn on_lex_build_error(&mut self, errors: Box<[LexBuildError]>);
+    fn missing_in_lexer(&mut self, missing: &HashSet<String>);
+    fn missing_in_parser(&mut self, missing: &HashSet<String>);
+    /// This function must return an `Err` variant if any of the following are true:
+    ///
+    /// - missing_in_lexer is called.
+    /// - missing_in_parser is called.
+    /// - on_lex_build_error is called.
+    fn results(&self) -> Result<(), Box<dyn Error>>;
```

Comment:
The `Result<(), Box<dyn Error>>` return value isn't currently very useful,
It's only called in cases it *must* return an `Err`, and since the `on_*` is never called when it is allowed to return `Ok`
It never has anything to do further analysis on, so it can't currently introduce it's own errors above and beyond the current error checking.  Perhaps we could consider adding a `fn analyse(...)`, It could be a method provided by the trait with an empty implementation even, which trait implementations can override.

Finally a call to `results()` at the end even if nothing `lrpar` considers an error has occurred.

Reply:
I'm not certain that I understand what this function does and why it does something distinct in the API. Is it because the other methods are expected to do some sort of mutation?

Reply:
Yes exactly, `self` is expected to basically collect the previous calls into some type `T` fully contained within self.
The trait exists to basically hide that `T` from lrpar.

In the `ariadne` trait implementation this is a vector of `ariadne::Report`s, in the simpler test impl, it is a `String` and the `result()` function calls `into()` on it.  for Sarif i'm expecting it to be a type that will be serializable to that format.

It is IMO not the prettiest thing, but I was running short on nicer ideas.

Reply:
My first thought was that `results` might want to take `self` (rather than `&self`) to make it clear that it's the "last" thing. But I'm not quite sure that's possible with Rust's type system in this case.

But then I wondered: do we even need such a "final" function? Can't whoever implements this trait summarise things on their own, without this? In other words, if they want to track whether there are any warnings/errors they can have (say) a `bool=false` and set it to `true` at the start of each implementation of a method in this trait?

Reply:
Part of the key to things like `Safir` output is that self is not consumed by this call.
my hope was that we could introduce a call like `dry_run`, which mirrors `build()` function but doesn't perform the filesystem operations, from which can pick up the results.  If you look at the last line in this test, it printlns some stuff out of the `ErrorHandler`, *after* calls to `build()`.

https://github.com/ratmice/error_callbacks_test/blob/implement_as_trait/build.rs#L287

We could definitely get away without `result`, but it means things like `build.rs` need to have some code which reaches into the type and prints warnings and errors, rather than only needing to do so in special cases like where you want to convert it to a specific format.  That was my reasoning for adding `results()`, and having it take a reference.

Reply:
> We could definitely get away without result, but it means things like build.rs need to have some code which reaches into the type and prints warnings and errors

I think I see where you're going with this. Can we unpack it into two questions?

1. What do we do by default as warnings/errors are discovered as things are processed?
2. What do we do by default at the final compile-time-`build` stage of a lexer/parser?

Those are two subtly different things, and I hadn't really thought about that before.

I think we still want the compile-time-`build` (for lexers/parsers) to produce an `Err` if there are errors discovered. That would remove a normal `build.rs`-using user from having to explicitly add checks for errors/warnings.

That then makes me wonder if we want (at least in part) the `ErrorHandler` traits to be outside the `ctbuilder` modules? In other words, some errors/warnings can be experienced by, say, an LSP, even though it won't have used `ctbuilder` --- and I *think* that there are some additional errors that only make sense in the "build lexer + parser together at compile-time" stage? Maybe one way of doing this is super/sub traits? e.g. for parsers maybe we'd have something like:

```rust
trait GrammarErrorHandler {
  // All the errors/warnings that a grammar can have whether it's compile-time or run-time
}

trait CompileTimeGrammarErrorHandler: GrammarErrorHandler {
  // Additional errors/warnings that make sense only at compile-time
}
```

and then `CTParserBuilder` would take an instance of `CompileTimeGrammarErrorHandler` and then (I think) `YaccGrammar` would take an instance of `GrammarErrorHandler`? I appreciate this implies a fairly deep change to how we handle all this stuff!

Reply:
That is an interesting question, I had debated whether I should actually put the current `GrammarErrorHandler` in `ctbuilder`, as in it's current form it *could* easily be used by `nimbeparse` outside `ctbuilder`, if do similar things for `RTParserBuilder`.

If i might add another layer here:

1. Errors/Warnings that happen during construction.
2. Errors/Warnings that can only be checked after construction of a (valid or invalid) `GrammarAST`  or successful construction of a `YaccGrammar` but are shared by `CT` and `RT`.
3. Errors that are `CT` specific. I think that this class is actually just filesystem errors, for reading the `.y` and writing the `.y.rs` (Which this trait isn't actually currently handling)

So I *believe* the current `GrammarErrorHandler` is the same as the compile-time or run-time one you mention above.

I'm not super well versed in this kind of trait magic, I actually tried to do something like this previously and ran into:
https://rust-lang.github.io/rfcs/3324-dyn-upcasting.html

That is being stablized just now though https://github.com/rust-lang/rust/pull/118133

>  I appreciate this implies a fairly deep change to how we handle all this stuff!

I'll have to remind myself how the construction of `Duplicate*Error` is produced, whether that is by buffering errors, or based on rules etc.  How deep of a change it is seems like it depends on that.

Reply:
About `results()`, and whether we need it,

> But then I wondered: do we even need such a "final" function?

I was thinking about this, and one thing that we could do is turn the `on_*` functions to returning `Box<dyn Err>` (back as they were in the callbacks branch).  We could still pass `&mut self`, but that mutation would only be needed if the trait impl wants to store errors.

The caller would be responsible for merging errors in perhaps a slightly awkward fashion though,
The simplest form of which could even be something like: 

```
    use std::error::Error;
    let x: Box<dyn Error> = "foo\n".into();
    let y: Box<dyn Error> = "bar\n".into();
    let z: Box<dyn Error> =  [x.to_string(), y.to_string()].join("Error:\n").into();
```

This sort of removes some of the strangeness about mutation, but maybe moves it elsewhere to anywhere we want to merge individual errors.

Reply:
>  The caller would be responsible for merging errors in perhaps a slightly awkward fashion

I've been chewing on this. I think it goes back to what I was asking in https://github.com/softdevteam/grmtools/pull/428#issuecomment-1858760041 about incremental/batch. The more I think about it, the more I think the API you're proposing is (correctly IMHO!) enabling incremental --- but there's nothing stopping a user "downgrading" it to batch. In other words, we're (potentially) telling the API user "as soon" as issues are discovered and they can then report those on immediately or later as they wish. In that sense, `results` is us trying to make the API support both incremental and batch natively --- but a batch user already has all the bits they need to make a batch API without `results`. So I'm inclined to keep the API minimal-ish and not have `results`.

Reply:
I tend to agree, I also think it simplifies the contract between caller and trait impl down to basically the method signature,
making the trait impl's more obviously correct.

The method which doesn't quite fit this mold of returning `Box<dyn Error>` is `on_grammar_warning()` I'd be inclined to just have it return `String` -- for which there is `From<String> for Box<dyn Error>`, so callers can still turn them into errors if they desire, we could even drop `warnings_are_errors` from the trait I believe, unless we want this for switching between "Warning" and "Error" in the text.

Reply:
Maybe a silly question -- but why do any of these questions have to return anything? AIUI we're feeding an external implementation with information about what's wrong, and I don't think we care about them filtering out anything? So in that sense (assuming we drop `results`) can't we just have each function take arguments and return nothing?

Reply:
I think it is pretty necessary to return something, because e.g. `build.rs` scripts are going to end in `.build()?`, and if we don't return an error variant, it seems wrong to return `Ok` even if we cannot proceed within the `build` process, due to e.g. the lack of a `YaccGrammar` value.

Reply:
I wonder if I've misunderstood. I thought the trait was something that *we* call and someone *outside* implements? In the case of (say) `CTBuilder` it can track (by implementing the trait) whether there were errors and then cause `build` to fail. Maybe I've got this backwards!

Reply:
Hmm, you've got it right, we call it and a `CTBuilder` calls it, and there can be both an *outside* implementation, and *default* implementation within the crate.

The main thing it is doing is delegating the responsibility of turning concrete e.g. `YaccGrammarError` types,
into the abstract `Box<dyn Error>`, the return type of `build()`  (Currently this is handled with custom formatting internally).  It still definitely tracks and knows which errors are fatal to to the build process.  It just acts as a shim for configuring the formatting, and if need be collecting the errors that occurred into values that escape `build()`.

If understanding your confusion, I do think it isn't totally clear whether this type should be driving any form of `ControlFlow` for error handling. I think part of this is a naming thing `ErrorHandler` implies it should, perhaps
`ErrorFormatter` would be better.  `warnings_are_errors` also seems to imply it should, but it is currently just treated
as a flag for whether we should return a `Err()` or `Ok()` from `results()` in the case where there are no errors but there are warnings.

Putting the `Box<dyn Error>` return value `on_grammar_error` and friends return values, puts control-flow squarely in the hands of the callers, and steps away from trying to mirror the built-in control-flow within the trait.

While leaving the `results()` call in kind of behaves like the trait is actively participating in the control-flow with the caller e.g. `CTBuilder`.  I do think it's definitely an issue we want to deliberately decide which side of the fence we stand on.
(and one that my current trait isn't particularly doing well at in the slightest!).

Reply:
> The main thing it is doing is delegating the responsibility of turning concrete e.g. YaccGrammarError types, into the abstract Box<dyn Error>, the return type of build()

Ah, so I got the wrong end of the stick (as is not uncommon!). So is what you're proposing a sort-of `ErrorMap` trait in the sense of "we give you, the external implementer an error, you `map` (in the sense of `iter().map(|| ...)` it into whatever you want, then we bundle all of those mapped things up and return them"?

Reply:
Yeah, that is a good way of putting it.  I lean towards that being the extent of what it should be doing since trying to do more seems just seems to make things complicated.

Reply:
Right, got it!

So what I'll do is explain what I *thought* you were proposing, then I'll go away and chew over which I think might be best!

What I thought was going to happen is that we provide this trait and as we find a warning/error we call the appropriate function. The user can then do various things: they can, for example, print things out, collate them to do something with later, or whatever. But as soon as we've called that function, our responsibility for that particular warning/error is over (i.e. we don't collate them).

So, for example, `CTBuilder` might choose to collate everything, and then print it out at when `build` is called. In contrast, an editor might try displaying popups immediately to a user while analysis (etc.) continues in a background thread.

Reply:
Yeah, originally it was kind of trying to do *both* things, e.g. it has to sort of provide the `map` because `build()` has to return *something*, in theory it could just return a token indicating `ErrorsOccurred`, and beyond that expect the trait to do what it does.  But also it has to dectect `cargo` and print things using a cargo in-band [cargo:warning](https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargo-warning).  So if we were to delegate all that to the trait, it is a bit of build script environment variable testing and special casing that each trait needs to implement when called from `build()`.

The above is what probably kept me from further going down that path.

Reply:
One last thing about e.g. pop-up alerts, I would note that this is still possible with the `map` centric approach, but it must act in *batches*, and require the *outside* to implement the pop-up behavior.
```
struct MyErrorThing {
   errors: Vec<YaccGrammarError>,
   warnings: Vec<YaccGrammarWarning>,
}

impl ErrorMapper for MyErrors {
     fn on_grammar_error(&mut self, error: YaccGrammarError) -> Box<dyn Error> {
         self.errors.push(error);
         // this perhaps would need to Box::new()?
         error.into() 
     }
}

then from build.rs:
```
let _ = ct.error_mapper(myerrors).build();
for e in my_errors {
    popup_alert(e);
}
```

Something like that anyways.